src/main/java/foundation/privacybydesign/email/ratelimit/MemoryRateLimit.java

```diff
@@ -0,0 +1,165 @@
+
+    private final Map<String, Long> ipLimits;
+    private final Map<String, Limit> emailLimits;
+
+    public MemoryRateLimit() {
+        ipLimits = new ConcurrentHashMap<>();
+        emailLimits = new ConcurrentHashMap<>();
+    }
+
+    public static MemoryRateLimit getInstance() {
+        if (instance == null) {
+            instance = new MemoryRateLimit();
+        }
+        return instance;
+    }
+
+    private long startLimitIP(long now) {
+        return now - IP_TIMEOUT*IP_TRIES;
+    }
+
+    @Override
+    protected synchronized long nextTryIP(String ip, long now) {
+        // Allow at most 1 try in each period (TIMEOUT), but kick in only
+        // after 3 tries. Thus while the user can do only 1 try per period
+        // over longer periods, the initial budget is 3 periods.
+        long limit = 0; // First try - last try was "long in the past".
+        if (ipLimits.containsKey(ip)) {
+            // Ah, there was a request before.
+            limit = ipLimits.get(ip);
+        }
+
+        long startLimit = startLimitIP(now);
+        if (limit < startLimit) {
+            // First visit or previous visit was long ago.
+            // Act like the last try was 3 periods ago.
+            limit = startLimit;
+        }
+
+        // Add a period to the current limit.
+        limit += IP_TIMEOUT;
+        return limit;
+    }
+
+    @Override
+    protected synchronized void countIP(String ip, long now) {
+        long nextTry = nextTryIP(ip, now);
+        if (nextTry > now) {
+            throw new IllegalStateException("counting rate limit while over the limit");
+        }
+        ipLimits.put(ip, nextTry);
+    }
+
+    // Is the user over the rate limit per e-mail address?
+    @Override
+    protected synchronized long nextTryEmail(String email, long now) {
+        // Rate limiter durations (sort-of logarithmic):
+        // 1   10 second
+        // 2   5 minute
+        // 3   1 hour
+        // 4   24 hour
+        // 5+  1 per day
+        // Keep log 5 days for proper limiting.
+
+        Limit limit = emailLimits.get(email);
+        if (limit == null) {
+            limit = new Limit(now);
+            emailLimits.put(email, limit);
+        }
+        long nextTry; // timestamp when the next request is allowed
+        switch (limit.tries) {
+            case 0: // try 1: always succeeds
+                nextTry = limit.timestamp;
+                break;
+            case 1: // try 2: allowed after 10 seconds
+                nextTry = limit.timestamp + 10 * SECOND;
+                break;
+            case 2: // try 3: allowed after 5 minutes
+                nextTry = limit.timestamp + 5 * MINUTE;
+                break;
+            case 3: // try 4: allowed after 3 hours
+                nextTry = limit.timestamp + 3 * HOUR;
+                break;
+            case 4: // try 5: allowed after 24 hours
+                nextTry = limit.timestamp + 24 * HOUR;
+                break;
```

Comment:
In the SMS issuer we saw that the 24h sometimes leads to issues when people are struggling and therefore request it multiple times. Since there is no cost factor involved here, can't we be a little bit more tolerable?

Reply:
Good one, made it like this:
attempt 1: no limit
attempt 2: 10 seconds
attempt 3: 1 minute
attempt 4: 5 minutes
attempt 5: 1 hour
attempt 6: 12 hours
attempt 6+ 2 per day